### ISSUE TITLE
Enable auto-scaling for web and streaming API

### ISSUE THREAD

**Issue by andreaswittig** [widdix/mastodon-on-aws#1]
Evaluate and implement auto-scaling for ECS services web and streaming API.

**Comment by scrappydog** [comment]
+1 for this feature

(some documentation of best practices on manual scale up process would be nice too)

**Comment by compuguy** [comment]
Based on several days of working with the three services, one can do an HA and auto-scaling configuration out of the box if one sets `AutoScaling` to true, and sets the `DesiredCount`, `MaxCapacity`, and `MinCapacity`. The only service that doesn't scale well is the sidekiq service. According to this page https://docs.joinmastodon.org/admin/scaling/#sidekiq, you can have multiple sidekiq services on different queues, except for the scheduler queue. There can only be one of those.

**Comment by scrappydog** [comment]
My Sidekiq task is regularly pegging at 100% CPU utilization... definitely need some guidance on configuring scaling...

**Comment by michaelwittig** [comment]
@scrappydog Same for us. I'm not sure if that is an issue. It likely doesn't matter if the background tasks utilize all resources as long as they finish withou much delay. For us, we see spikes to 100% but only for minutes. Do you see the same pattern?
<img width="1096" alt="Screenshot 2022-11-28 at 09 42 10" src="https://user-images.githubusercontent.com/950078/204618153-159351d1-ce24-4fa2-b6cd-d22249ecde2e.png">

**Comment by scrappydog** [comment]
I bumped the CPU allocation up on the Sidekiq task to CPU .5 vCPU | Memory 3 GB... 

This feels happier for now... but it doesn't address the real scalability question...

**Comment by michaelwittig** [comment]
I opened up #20 for sidekiq. This issue is about auto-scaling for web and streaming API.

Enabling auto-scaling is not the big deal here. What we need is a good metric to trigger scale out/in. And we need a test workload to test tis with. I have no idea how we can simulate mastodon load. If anyone here is reading this running an instance with enough users to benefit rom auto-scaling please let us know.

**Comment by nodomain** [comment]
Just add a relay server and you will have CPU load in a minute. 

https://github.com/brodi1/activitypub-relays

**Comment by compuguy** [comment]
> I opened up #20 for sidekiq. This issue is about auto-scaling for web and streaming API.
> 
> Enabling auto-scaling is not the big deal here. What we need is a good metric to trigger scale out/in. And we need a test workload to test tis with. I have no idea how we can simulate mastodon load. If anyone here is reading this running an instance with enough users to benefit rom auto-scaling please let us know.

Yeah it's quite easy to autoscale the web and streaming API's. But for most people it's #20 that's more important since Sidekiq does most of the heavy lifting for Mastodon...
